--- /sys/src/9/boot/boot.c	Sat Mar 28 13:11:48 2015
+++ ../frontmods/boot/boot.c	Fri Dec 22 02:26:47 2017
@@ -9,6 +9,7 @@
 {
 	char cputype[64];
 	char buf[32];
+	char bootcmd[64];
 
 	fmtinstall('r', errfmt);
 
@@ -36,6 +37,7 @@
 	USED(argc);
 
 	readfile("#e/cputype", cputype, sizeof(cputype));
+	readfile("#e/bootcmd", bootcmd, sizeof(bootcmd));
 
 	/* setup the boot namespace */
 	bind("/boot", "/bin", MAFTER);
@@ -44,5 +46,7 @@
 	snprint(buf, sizeof(buf), "/%s/bin", cputype);
 	bind(buf, "/bin", MAFTER);
 	bind("/rc/bin", "/bin", MAFTER);
+	if(strcmp(bootcmd, "plan9rc")==0)
+		exec("/boot/plan9rc", argv);
 	exec("/bin/bootrc", argv);
 }
