--- /sys/src/cmd/auth/factotum/fs.c	Sat Dec 31 23:31:18 2016
+++ ../frontmods/patched/factotumfs.c	Thu Dec 22 01:31:50 2016
@@ -12,6 +12,7 @@
 char		*service;
 int		sflag;
 int		uflag;
+int		nomntcs;
 
 extern Srv		fs;
 static void		notifyf(void*, char*);
@@ -105,6 +106,8 @@
 	case 'u':		/* user: set hostowner */
 		uflag = 1;
 		break;
+	case 'x':
+		nomntcs = 1;
 	default:
 		usage();
 	}ARGEND
