--- /sys/src/9/pc/pc	Mon Dec 18 01:49:42 2017
+++ ../frontmods/pc/pcf	Mon Dec 18 10:56:51 2017
@@ -1,8 +1,7 @@
-# pc - 386 pc terminal with local disk
+# pcf - pc terminal with local disk
 dev
 	root
 	cons
-	swap
 	arch
 	pnp		pci
 	env
@@ -10,6 +9,7 @@
 	proc
 	mnt
 	srv
+	zrv
 	shr
 	dup
 	rtc
@@ -20,7 +20,6 @@
 	fs
 
 	ether		netif
-	bridge		netif log
 	ip		arp chandial ip ipv6 ipaux iproute netlog nullmedium pktmedium ptclbsum386 inferno
 
 	draw		screen vga vgax swcursor
@@ -38,9 +37,6 @@
 	i82365		cis
 	uart
 	usb
-	
-	segment
-	vmx
 
 link
 	segdesc
@@ -85,7 +81,6 @@
 	usbuhci
 	usbohci
 	usbehci		usbehcipc
-	usbxhci		pci
 
 	audiosb16	dma
 	audioac97	audioac97mix
@@ -104,7 +99,6 @@
 	sdodin		pci sdscsi led
 	sdvirtio	pci sdscsi
 	sdmmc		pci pmmc
-	sdnvme		pci
 	sdloop
 
 	uarti8250
@@ -154,6 +148,7 @@
 
 bootdir
 	/$objtype/bin/paqfs
-	/$objtype/bin/auth/factotum
 	bootfs.paq
 	boot
+	/n/rootless/bootdir.extras/factotum
+	/n/rootless/bootdir.extras/plan9rc
