--- /sys/src/9/pc/pcf	Sun Jun 28 16:34:10 2015
+++ frontmods/pc/pcf	Sun Dec 18 06:55:33 2016
@@ -147,6 +147,16 @@
 
 bootdir
 	/$objtype/bin/paqfs
-	/$objtype/bin/auth/factotum
 	bootfs.paq
 	boot
+	/bin/bind
+	/bin/mount
+	/bin/read
+	/bin/cat
+	/bin/echo
+	/bin/test
+	/n/rootless/bootdir.extras/rc
+	/n/rootless/bootdir.extras/rcmain
+	/n/rootless/bootdir.extras/factotum
+	/n/rootless/bootdir.extras/bootpaq
+	/n/rootless/bootdir.extras/plan9rc
