--- /sys/src/9/pc64/pc64	Sat Dec 31 23:31:14 2016
+++ ../frontmods/pc64/pc64	Tue Jan  3 10:15:10 2017
@@ -65,7 +65,7 @@
 #	ethervt6105m	pci ethermii
 #	ethersink
 #	ethersmc	devi82365 cis
-	etheryuk	pci
+#	etheryuk	pci
 #	etherwavelan	wavelan devi82365 cis pci
 	etheriwl	pci wifi
 	etherwpi	pci wifi
@@ -145,6 +145,7 @@
 
 bootdir
 	/$objtype/bin/paqfs
-	/$objtype/bin/auth/factotum
 	bootfs.paq
 	boot
+	/n/rootless/bootdir.extras/factotum
+	/n/rootless/bootdir.extras/plan9rc
