--- /sys/src/9/port/proc.c	Sun Dec 17 12:32:57 2017
+++ ../frontmods/port/proc.c	Fri Dec 22 02:26:47 2017
@@ -621,6 +621,7 @@
 	p->egrp = nil;
 	p->fgrp = nil;
 	p->rgrp = nil;
+	p->sgrp = nil;
 	p->pdbg = nil;
 	p->fpstate = FPinit;
 	p->kp = 0;
@@ -1090,6 +1091,7 @@
 	Egrp *egrp;
 	Rgrp *rgrp;
 	Pgrp *pgrp;
+	Sgrp *sgrp;
 	Chan *dot;
 	void (*pt)(Proc*, int, vlong);
 
@@ -1119,6 +1121,13 @@
 		closeegrp(egrp);
 	if(rgrp != nil)
 		closergrp(rgrp);
+	/* sgrp is nilled out here because closefgrp may need srvclose */
+	qlock(&up->debug);
+	sgrp = up->sgrp;
+	up->sgrp = nil;
+	qunlock(&up->debug);
+	if(sgrp != nil)
+		closesgrp(sgrp);
 	if(dot != nil)
 		cclose(dot);
 	if(pgrp != nil)
